interrupt.go

```diff
@@ -0,0 +1,122 @@
+package terminal
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+
+	"fortio.org/log"
+)
+
+type InterruptReader struct {
+	reader  io.Reader // stdin typically
+	buf     []byte
+	bufSize int
+	err     error
+	mu      sync.Mutex
+	cancel  context.CancelFunc
+}
+
+var ErrInterrupted = errors.New("terminal interrupted")
+
+// NewInterruptReader creates a new interrupt reader.
+// it needs to be Start()ed to start reading from the underlying reader
+// and intercept Ctrl-C and listen for interrupt signals.
+func NewInterruptReader(reader io.Reader, bufSize int) *InterruptReader {
+	ir := &InterruptReader{
+		reader:  reader,
+		bufSize: bufSize,
+		buf:     make([]byte, 0, bufSize),
+	}
+	log.Config.GoroutineID = true
+	return ir
+}
+
+// Start or restart (after a cancel/interrupt) the interrupt reader.
+func (ir *InterruptReader) Start(ctx context.Context) (context.Context, context.CancelFunc) {
+	ir.mu.Lock()
+	defer ir.mu.Unlock()
+	if ir.cancel != nil {
+		ir.cancel()
+	}
+	nctx, cancel := context.WithCancel(ctx)
+	ir.cancel = cancel
+	go func() {
+		ir.start(nctx)
+	}()
+	return nctx, cancel
+}
+
+// Implement io.Reader interface.
+func (ir *InterruptReader) Read(p []byte) (int, error) {
+	ir.mu.Lock()
+	n := copy(p, ir.buf)
+	ir.buf = ir.buf[n:]
+	err := ir.err
+	ir.err = nil
+	ir.mu.Unlock()
+	return n, err
+}
+
+const CtrlC = 3
```

Comment:
Do you need this to be exported?

Reply:
not exactly but it also doesn't hurt and is nicer in the code than `3` or `ctrlC`